File: 2022/python/day13.py
from itertools import zip_longest


def correct(left, right):
    if left < right:
        return True
    if right < left:
        return False
    return None


def main():
    input_file = "data/day13.txt"
    with open(input_file) as f:
        lines = f.read()

    pairs = []

    for pair in lines.split("\n\n"):
        (l1, l2) = pair.split("\n")
        left = eval(l1)
        right = eval(l2)
        print("original:", left, right)
        result = test_pair(left, right)
        print(result)
        pairs.append(result)

    print(pairs)
    print("Part1: ", sum(i + 1 for i, result in enumerate(pairs) if result))


def test_pair(left, right):
    for l, r in zip_longest(left, right):
        if type(l) == list and len(l) == 0:
            return True
        if type(r) == list and len(r) == 0:
            return False

        l = 0 if l == None else l
        r = 0 if r == None else r
        to_compare = []
        to_compare.append((l, r))
        while to_compare:
            print("queue", to_compare)
            (compare_l, compare_r) = to_compare.pop(0)
            print("compare:", compare_l, compare_r)
            if type(compare_l) == int and type(compare_r) == int:
                pair_result = correct(compare_l, compare_r)
                if pair_result == None:
                    continue
                else:
                    return pair_result
            else:
                compare_l = [compare_l] if type(compare_l) == int else compare_l
                compare_r = [compare_r] if type(compare_r) == int else compare_r
                if len(compare_l) == 0:
                    return True
                if len(compare_r) == 0:
                    return False
                for l_inner, r_inner in zip_longest(compare_l, compare_r):
                    l_inner = 0 if l_inner == None else l_inner
                    r_inner = 0 if r_inner == None else r_inner
                    to_compare.append((l_inner, r_inner))
    return False


if __name__ == "__main__":
    main()

File: 2022/src/solutions/day13.rs
// // use itertools::Itertools;
// use std::{collections::VecDeque, fs::read_to_string, path::Path};

// use serde_json::{Value, Number};

// pub fn solution() {
//     let base = Path::new(env!("CARGO_MANIFEST_DIR")).join("data/example.txt");
//     // let mut data = String::new();
//     // reader.read_line(&mut data).expect("");
//     let file_str = read_to_string(base).expect("Could not read file contents");

//     let mut pair_results = Vec::new();
//     // let t = serde_json::number::Number;
//     let zero = Value::Number(PosInt(0));

//     // let zero_value = Value::Null;
//     'pairs: for pair in file_str.split("\n\n") {
//         let (l1, l2) = pair.split_once("\n").unwrap();
//         let left: Value = serde_json::from_str(&l1).unwrap();
//         let right: Value = serde_json::from_str(&l2).unwrap();

//         let left_len = left.as_array().unwrap().len();
//         let right_len = right.as_array().unwrap().len();

//         // println!("{:?}", left[1]);
//         // println!("{:?}", left[1].as_u64());
//         // println!("{:?}", left[1].as_array());

//         for i in 0..left_len {
//             let mut to_compare = VecDeque::new();
//             to_compare.push_back((left[i].clone(), right[i].clone()));
//             while let Some((left_item, right_item)) = to_compare.pop_front() {
//                 let result = match (left_item.as_u64(), right_item.as_u64()) {
//                     // Both are ints, compare them
//                     (Some(left_int), Some(right_int)) => correct(left_int, right_int),
//                     // Both are not ints, convert to array
//                     (_, _) => {
//                         let left_array = left_item.as_array().expect("Shit");
//                         let right_array = right_item.as_array().expect("Shit");
//                         for i in 0..left_array.len().max(right_array.len()) {
//                             to_compare.push_back((
//                                 left_array.get(i).unwrap_or(&Value::Null).clone(),
//                                 right_array.get(i).unwrap_or(&Value::Null).clone()
//                             ));
//                         };
//                         None
//                     }
//                 };
//                 if result.is_some() {
//                     pair_results.push(resdef main():
//     input_file = "data/day1.txt"
//     with open(input_file) as f:
//         weights = [
//             sum(map(int, weight_str.split())) for weight_str in f.read().split("\n\n")
//         ]
//         print(max(weights))


// if __name__ == "__main__":
//     main()
// ult.unwrap());
//                     continue 'pairs
//                 }
//             }
//         }
//     }
//     println!("{:?}", pair_results)
// }

// fn correct(left: u64, right: u64) -> Option<bool> {
//     if left < right {
//         return Some(true);
//     } else if left > right {
//         return Some(false);
//     }
//     None
// }

// // fn parse_items(list_str: &str) {
// //     let items = Vec::new();
// //     let inner_str = &list_str[1..list_str.len()-1];

// // }
